specs/mgt-search.md

```diff
@@ -0,0 +1,182 @@
+
+## Templates
+
+ `mgt-search` supports several [templates](../customize-components/templates.md) that you can use to replace certain parts of the component. To specify a template, include a `<template>` element inside a component and set the `data-type` value to one of the following.
+
+| Data type | Data context | Description |
+| --- | --- | --- |
+| loading | null: no data | The template used to render the state of picker while request to graph is being made. |
+| error | null: no data | The template used if user search returns no users. |
+| no-data | null: no data | An alternative template used in dropdown list if suggestion returns no any results |
+| flyout-header | null: no data | The template to render flyout header. |
+| flyout-footer | null: no data | The template to render flyout footer. |
+| suggestion-input | null: no data | The template to render search input box. |
+| suggested-people-header | null: no data | The template to render people entity label. |
+| suggested-query-header | null: no data | The template to render text entity label. |
+| suggested-file-header | null: no data | The template to render file entity label. |
+| customized-{other entity type}-header | null: no data | The template to render other entity label, for example customized-sample1-header |
+| suggested-people | suggestionPeople[]: The people suggestion details  | The template to render people entity. |
+| suggested-query | suggestionQuery[]: The text suggestion details  | The template to render text entity. |
+| suggested-file | suggestionFile[]: The file suggestion details  | The template to render file entity. |
+| customized-{other entity type} | any[]: Customizing suggestion details  | The template to render customized entity, for example customized-sample1 |
+
+The following examples shows how to use customized template `suggested-query`.
+
+```html
+<mgt-search>
+    <template data-type="suggested-query">
+        <div>{{query}}</div>
+    </template>
+</mgt-search>
+```
+
+
+## Customized
+Customized entity types must be supported by microsoft graph suggestion API.
+```html
+<mgt-search entity-types="query,file,people,sample1">
+    <template data-type="customized-sample1-header">
+        <div>header hahaha</div>
+    </template>
+
+    <template data-type="customized-sample1">
+        <div>{{this}}</div>
+    </template>
+</mgt-search>
+```
+
+## Proposed Solution
+
+## Example
+
+### Example 1: basic usage
+```<mgt-search></mgt-search>```
+
+### Example 2: get suggestion data and define click and enter key action by adding a listener
+```
+<mgt-search> </mgt-search>
+
+<script>
+
+document.querySelector('mgt-search').addEventListener('onEntityClick', e => {
+    console.log(e.detail);
+});
+
+document.querySelector('mgt-search').addEventListener('onEnterPress', e => {
+  console.log(e.detail);
+});
+
+</script>
+
+```
+
+### Example4: Select entity types & Sort & Suggestion Amount
```

Comment:
This seems a bit complex? I wonder if we could instead add a property to set the max suggestions per type?

ex:

```html
<mgt-search entity-types="query, file, people" max-suggestions="3"></mgt-search>
```

Reply:
I have a concern here, if we support customize entity type, Developers should have ability to customize the count of each entity's suggestion count. if I set a fixed parameters, **for example max-file-suggestion = "3",** once developer add new entity type **entitySample**, **Developers don't have a way to customize the max-entitySample-suggestion.**

and in current design **"query, file,  people"**, all of  them have the default value, which equals to "query-3, file-3,  people-3", only when developer want to customize the suggestion count, they can use such like  "query, file-10,  people"

**I am not sure if there are more suitable ways to do it, If I change the design here, there are some different solutions:**
1: As  the comment here, all of this entity should have same suggestion numbers.
        max-suggestions="3"

2: Add 3 new parameters to allow developer change the query, file, people suggestion count. 
customized component should use default value

3: Add 3 new parameters to allow developer change the query, file, people suggestion count. 
customized component should use any parameter to control the suggestion count.

Reply:
I agree with this. Honestly, I'd prefer a simple solution to start with and add features as we get feedback. Keeping a single property that will set the same max number to all suggestions makes sense.